feat: Add post "securing Kubewarden policies" (#82)
kubewarden/kubewarden.io: content/blog/2022/04/securing-kubewarden-policies.md

---
title: Secure Supply Chain with Kubewarden: securing Kubewarden policies
authors:
- Víctor Cuadrado Juan
date: 2022-04-21
lastmod: 2022-04-21
---

With the recent releases, Kubewarden stack is now gaining support for the
[Sigstore standard](https://www.sigstore.dev/).

In this post, we focus on Kubewarden Policies and how to create a Secure Supply
Chain for them.

## Sigstore?

Since a full Sigstore dive is not for this post, we recommend [their nice docs](https://docs.sigstore.dev).

In short, Sigstore provides an automatable workflow thought out to match the
distributed Open Source development model. The workflow specifies how to
digitally sign and verify artifacts; in our case, OCI images. It also provides a
transparency log to monitor such signatures. The workflow allows to sign
artifacts with traditional Public-Private key pairs, or sign in Keyless mode.

> Keyless mode is interesting: it uses short-lived CA certs with you as issuer, and

Author: Indeed, that's wrong!  Reworded in https://github.com/kubewarden/kubewarden.io/pull/82/commits/332d4a39ed07468e8fb68e895ed4792405f8373c.

the signatures.

The Public-Private key pair signing is straightforward, using [sigstore/cosign](https://github.com/SigStore/cosign):

```console
$ COSIGN_PASSWORD=yourpass cosign generate-key-pair

Private key written to cosign.key
Public key written to cosign.pub

$ COSIGN_PASSWORD=yourpass cosign sign \
  --key cosign.key --annotations blog=yes \
  ghcr.io/kubewarden/policies/user-group-psp:v0.2.0

Pushing signature to: ghcr.io/kubewarden/policies/user-group-psp
```

The Keyless mode is more interesting:
```console
$ COSIGN_EXPERIMENTAL=1 cosign sign \
  --annotations blog=yes \
  ghcr.io/kubewarden/policies/user-group-psp:v0.2.0

Generating ephemeral keys...
Retrieving signed certificate...
Your browser will now be opened to:
https://oauth2.sigstore.dev/auth/auth?access_type=online&client_id=sigstore&code_challenge=(...)
Successfully verified SCT...
tlog entry created with index: (...)
Pushing signature to: ghcr.io/viccuad/policies/volumes-psp
```

What happened? `cosign` prompted us for an OpenID Connect provider on the
browser, which authenticated us, and allowed Fulcio to sign our ephemeral keys.

> If this were to happen in a CI, for example GH Actions, the CI would provide an authentication token.

Reviewer: we can mention that this works in our GH action because cosign has support for detecting some automated environments and producing an identity token. It currently supports github and google. If they want to use a different CI users can provide an `--indentity-token`. more info https://github.com/sigstore/cosign/blob/main/KEYLESS.md#identity-tokens

Author: That sounds better, yes. Done in https://github.com/kubewarden/kubewarden.io/pull/82/commits/ce239ddb36369430df124377bc535a8f4d9ecf42.

    repo: ~
    annotations: ~
anyOf: ~
```

The verification config format has several niceties, [see its reference docs](TODO point to VerificationConfig reference in docs).
For example, `kind: githubAction` with `owner` and `repo`, instead of checking the `issuer` and
`subject` strings blindly. Or `anyOf` a list of signatures, with `anyOf.atLeast` a number
of them: this allows for accepting at least a specific number of signatures, and
makes migration between signatures in your cluster easy. It's the little things :).

If you want support for other CIs (such as Gitlab, Jenkins, etc) drop us some
words!

Once you have crafted your verification config, create your ConfigMap:
```console
$ kubectl create configmap my-verification-config \
  --from-file=verification-config=./my-verification-config.yml \
  --namespace=kubewarden
```

And pass it to your PolicyServers in `spec.VerificationConfig`, or if using the
default PolicyServer from the `kubewarden-defaults` chart, set it there with for example:
```console
$ helm upgrade --set policyServer.verificationConfig=my-verification-config \
  --wait -n kubewarden kubewarden-defaults ./kubewarden-defaults
```

## Recap

> Using `cosign`, `kwctl verify`, and  `kwctl inspect`, we can sign our authored

Reviewer: we use `cosign` to sign, and `kwctl verify` to verify.

> Using `cosign`, `kwctl verify`, and  `kwctl inspect`, we can sign our authored

In this sentence it looks to me that you need these 3 tools to sign

Author: True! done in https://github.com/kubewarden/kubewarden.io/pull/82/commits/29edb820d16ba573bdd80b698087c17b69102e25

---
title: Secure Supply Chain with Kubewarden: securing Kubewarden policies
authors:
- Víctor Cuadrado Juan
date: 2022-04-21
lastmod: 2022-04-21
---

With the recent releases, Kubewarden stack is now gaining support for the
[Sigstore standard](https://www.sigstore.dev/).

In this post, we focus on Kubewarden Policies and how to create a Secure Supply
Chain for them.

## Sigstore?

Since a full Sigstore dive is not for this post, we recommend [their nice docs](https://docs.sigstore.dev).

In short, Sigstore provides an automatable workflow thought out to match the
distributed Open Source development model. The workflow specifies how to
digitally sign and verify artifacts; in our case, OCI images. It also provides a
transparency log to monitor such signatures. The workflow allows to sign
artifacts with traditional Public-Private key pairs, or sign in Keyless mode.

> Keyless mode is interesting: signatures are created with short-lived CA certs

Reviewer: ```suggestion
Keyless mode is interesting: signatures are created with short-lived CA certs
```
```suggestion
Keyless mode is interesting: signatures are created with short-lived certs
```

> Keyless mode is interesting: signatures are created with short-lived CA certs

Reviewer: ```suggestion
Keyless mode is interesting: signatures are created with short-lived CA certs
```
```suggestion
Keyless mode is interesting: signatures are created with short-lived certs
```

Reviewer: I think the short-lived certs are not CA

with an OpenID Connect (OIDC) service as issuer, and with Sigstore's CA

> ([Fulcio](https://github.com/sigstore/fulcio)) as root CA.

Author: No, thanks for the corrections! resolving this conversation, see the other conversation that aglutinates these 2 paragraphs.

The OpenID Connect service (SSO via your own Okta instance, Github, Google, etc)
proves to Fulcio that you are who you say you are, and creates the short-lived
certificate. This certificate has the OIDC service as issuer, and info related
to your artifact and how it was build in the subject key. And with it, the

> signature is created.

Author: True! I've reworded the 2 paragraphs in a commit: https://github.com/kubewarden/kubewarden.io/pull/82/commits/5c084a2ad992b4f8b7a96c03eeda0e40b6aeaf03.

Please tell me how it sounds now.

https://oauth2.sigstore.dev/auth/auth?access_type=online&client_id=sigstore&code_challenge=(...)
Successfully verified SCT...
tlog entry created with index: (...)
Pushing signature to: ghcr.io/viccuad/policies/volumes-psp
```

What happened? `cosign` prompted us for an OpenID Connect provider on the
browser, which authenticated us, and allowed Fulcio to sign our ephemeral keys.

If this were to happen in a CI, the CI can provide an OIDC identity token in
their environment. `cosign` has support for detecting some automated
environments and producing an identity token. Currently that covers
[Github And Google Cloud, but one can always use a flag](https://github.com/sigstore/cosign/blob/main/KEYLESS.md#identity-tokens).

This is how it works for policies built by the Kubewarden team in GH Actions:
we [call cosign](https://github.com/kubewarden/github-actions/blob/520eaa5e479fcb253ba09009c63f7fcfca1f743d/policy-release/action.yaml#L43),
and the resulting signature will be signed by a cert. This cert has:
- An `issuer`,  telling you who certifies the image:
  `https://token.actions.githubusercontent.com`
- A `subject` related to the specific workflow and worker, for example:

>   `https://github.com/kubewarden/policy-secure-pod-images/.github/workflows/release.yml@refs/heads/main`

Author: Thanks! applied and added command in https://github.com/kubewarden/kubewarden.io/pull/82/commits/1d74424771a66a0e2be4a62d39e921c840a50be2.

default PolicyServer from the `kubewarden-defaults` chart, set it there with for example:
```console
$ helm upgrade --set policyServer.verificationConfig=my-verification-config \
  --wait -n kubewarden kubewarden-defaults ./kubewarden-defaults
```

## Recap

Using `cosign sign` we can sign our authored policies. With `kwctl verify`, we
can verify them, and with  `kwctl inspect` (and other tools such as `crane
manifest`), we can inspect the signatures.
We can keep using `kwctl pull` and `kwctl run` to test policies locally as in
the past, plus now verify their signatures too. Once we are satisfied,
[we can deploy Kubewarden PolicyServers so they enforce those signatures](TODO link to doc page on verification). And if we want, the
same verification config format can be used for `kwctl` and the cluster stack.

This way we are sure that the policies come from their stated authors, and have
not been tampered with. Phew!

We the Kubewarden team are curious on how you approach this. What workflows are you interested in?
What challenges do you have? Drop as word in our [Slack channel](https://kubernetes.slack.com/archives/C01T3GTC3L7)

> or in a [Github issue](https://github.com/kubewarden)!

Reviewer 2: ```suggestion
or in a [GitHub issue](https://github.com/kubewarden)!
```

If you want to verify policies locally, you now can use `kwctl verify`:
```console
$ kwctl verify --github-owner kubewarden registry://ghcr.io/kubewarden/policies/pod-privileged:v0.1.10
$ echo $?
0
```

And when testing policies locally with `kwctl pull` or `kwctl run`, you can also
enable signature verification by using any verification related flag. For example:
```console
$ kwctl pull --github-owner kubewarden registry://ghcr.io/kubewarden/policies/pod-privileged:v0.1.10
$ echo $?
0
```

All the policies from the Kubewarden team are signed in keyless mode by the
workers of the CI job. In our case, the workers used in the CI job of Github.
We don't leave certs around, and they are verifiable by third parties.

### Enforcing signature verification for instantiated Kubewarden policies

You can now configure
[PolicyServers](https://doc.crds.dev/github.com/kubewarden/kubewarden-controller/policies.kubewarden.io/PolicyServer/v1alpha2)
to enforce that all policies being run need to be signed. When
deploying Kubewarden via Helm charts, you can do it so for the default
PolicyServer installed by `kubewarden-defaults` chart.

> For this, the PolicyServers have a new [`spec.VerificationConfig`](https://doc.crds.dev/github.com/kubewarden/kubewarden-controller/policies.kubewarden.io/PolicyServer/v1alpha2#spec-verificationConfig) argument.

Reviewer 2: ```suggestion
For this, the PolicyServers has a new [`spec.VerificationConfig`](https://doc.crds.dev/github.com/kubewarden/kubewarden-controller/policies.kubewarden.io/PolicyServer/v1alpha2#spec-verificationConfig) argument.
```

Here, you can put the name of a ConfigMap containing a "verification config", to
specify the needed signatures.

You can obtain a default verification config for policies from the Kubewarden
team with:

```yaml
$ kwctl scaffold verification-config
# Default Kubewarden verification config
#
# With this config, the only valid policies are those signed by Kubewarden
# infrastructure.
#
# This config can be saved to its default location (for this OS) with:
#   kwctl scaffold verification-config > /home/youruser/.config/kubewarden/verification-config.yml
#
# Providing a config in the default location enables Sigstore verification.
# See https://docs.kubewarden.io for more Sigstore verification options.
---
apiVersion: v1
allOf:
  - kind: githubAction
    owner: kubewarden
    repo: ~
    annotations: ~
anyOf: ~
```

> The verification config format has several niceties, [see its reference docs](TODO point to VerificationConfig reference in docs).

Reviewer 2: REMINDER: put the link

Author: Done in https://github.com/kubewarden/kubewarden.io/pull/82/commits/0093c152ae5b278f4cc41cc7c690f3ea7745d44c.

For example, `kind: githubAction` with `owner` and `repo`, instead of checking the `issuer` and
`subject` strings blindly. Or `anyOf` a list of signatures, with `anyOf.atLeast` a number
of them: this allows for accepting at least a specific number of signatures, and
makes migration between signatures in your cluster easy. It's the little things 🤓.

If you want support for other CIs (such as GitLab, Jenkins, etc) drop us some
words!

Once you have crafted your verification config, create your ConfigMap:
```console
$ kubectl create configmap my-verification-config \
  --from-file=verification-config=./my-verification-config.yml \
  --namespace=kubewarden
```

And pass it to your PolicyServers in `spec.VerificationConfig`, or if using the
default PolicyServer from the `kubewarden-defaults` chart, set it there with for example:
```console
$ helm upgrade --set policyServer.verificationConfig=my-verification-config \
  --wait -n kubewarden kubewarden-defaults ./kubewarden-defaults
```

## Recap

Using `cosign sign` policy authors can sign our author their policies. All the policies owned by the Kubewarden

> team have been signed in this way.

Reviewer 2: ```suggestion
team have already been signed in this way.
```

With `kwctl verify`, operators

> can verify them, and with  `kwctl inspect` (and other tools such as `crane

Reviewer 2: ```suggestion
can verify them, and with `kwctl inspect` (and other tools such as `crane
````

manifest`), operators can inspect the signatures.
We can keep using `kwctl pull` and `kwctl run` to test policies locally as in
the past, plus now verify their signatures too. Once we are satisfied,

> [we can deploy Kubewarden PolicyServers so they enforce those signatures](TODO link to doc page on verification). And if we want, the

Reviewer 2: REMINDER: add the final link

Author: Done in https://github.com/kubewarden/kubewarden.io/pull/82/commits/0093c152ae5b278f4cc41cc7c690f3ea7745d44c.

same verification config format can be used for `kwctl` and the cluster stack.

This way we are sure that the policies come from their stated authors, and have
not been tampered with. Phew!

We the Kubewarden team are curious on how you approach this. What workflows are you interested in?

> What challenges do you have? Drop as word in our [Slack channel](https://kubernetes.slack.com/archives/C01T3GTC3L7)

Reviewer 2: ```suggestion
What challenges do you have? Drop us a word in our [Slack channel](https://kubernetes.slack.com/archives/C01T3GTC3L7)
```
